plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3ExchangeStorage.java

```diff
@@ -0,0 +1,686 @@
+            regionOrEndpointSet = true;
+        }
+
+        if (!regionOrEndpointSet) {
+            clientBuilder.withRegion(US_EAST_1);
+            clientBuilder.setForceGlobalBucketAccessEnabled(true);
+        }
+
+        return clientBuilder.build();
+    }
+
+    private static boolean isDirectory(URI uri)
+    {
+        return uri.toString().endsWith(PATH_SEPARATOR);
+    }
+
+    private static class S3ExchangeSliceOutput
+            extends SliceOutput
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(S3ExchangeSliceOutput.class).instanceSize();
+
+        private final AmazonS3 s3;
+        private final String bucketName;
+        private final String key;
+        private final Supplier<String> uploadIdFactory;
+        private final ExecutorService uploadExecutor;
+        private final Optional<SSECustomerKey> sseCustomerKey;
+
+        private int currentPartNumber;
+        private byte[] buffer;
+        private int bufferSize;
+
+        private boolean failed;
+        private Optional<String> uploadId = Optional.empty();
+        private Future<UploadPartResult> inProgressUploadFuture;
+        private final List<UploadPartResult> parts = new ArrayList<>();
+
+        public S3ExchangeSliceOutput(
+                AmazonS3 s3,
+                String bucketName,
+                String key,
+                Supplier<String> uploadIdFactory,
+                ExecutorService uploadExecutor,
+                int partSize,
+                Optional<SSECustomerKey> sseCustomerKey)
+        {
+            this.s3 = requireNonNull(s3, "s3 is null");
+
+            this.buffer = new byte[partSize];
+
+            this.bucketName = requireNonNull(bucketName, "bucketName is null");
+            this.key = requireNonNull(key, "key is null");
+            this.uploadIdFactory = requireNonNull(uploadIdFactory, "uploadIdFactory is null");
+            this.uploadExecutor = requireNonNull(uploadExecutor, "uploadExecutor is null");
+            this.sseCustomerKey = requireNonNull(sseCustomerKey, "sseCustomerKey is null");
+        }
+
+        @Override
+        public void writeInt(int value)
+        {
+            writeBytes(ByteBuffer.allocate(Integer.BYTES).order(ByteOrder.LITTLE_ENDIAN).putInt(value).array());
+        }
+
+        @Override
+        public void writeBytes(Slice slice)
+        {
+            writeBytes(slice.getBytes());
+        }
+
+        @Override
+        public void writeBytes(byte[] buffer)
+        {
+            writeBytes(buffer, 0, buffer.length);
+        }
+
+        @Override
+        public void writeBytes(byte[] bytes, int offset, int length)
+        {
+            while (length > 0) {
+                int copied = min(buffer.length - bufferSize, length);
+                arraycopy(bytes, offset, buffer, bufferSize, copied);
+                bufferSize += copied;
+
+                try {
+                    flushBuffer(false);
```

Comment:
why flushing always. Would it make sense to only flush when we run out of place in `buffer`?

Reply:
Oh actually I see that `flushBuffer` does not always flush. Only when buffer is full. Maybe rename the method to `flushBufferIfNeeded` so it is less confusing.

Reply:
Same as above, trying to be consistent with `TrinoS3FileSystem`. I agree with your suggestion, though.